doc/designs/fast-param-find.md

```diff
@@ -0,0 +1,326 @@
+            'access' => 'writeonly',
+            'setaction' => qq(
+                if (!scrypt_set_membuf(&ctx->pass, &ctx->pass_len, p))
+                    return 0;
+            ),
+        ),
+
+        'KDF_PARAM_SALT' => (
+            'type' => 'octet string',
+            'access' => 'readwrite',
+            'setaction' => qq(
+                if (!scrypt_set_membuf(&ctx->salt, &ctx->salt_len, p))
+                    return 0;
+            ),
+            'getaction' => qq(
+                p->return_size = ctx->salt_len;
+                if (p->data_size >= ctx->salt_len)
+                    memcpy(p->data, ctx->salt, p->data_size >= ctx->salt_len);
+            ),
+        ),
+
+        'KDF_PARAM_SCRYPT_N' => (
+            'type' => 'integer',
+            'ctype' => 'uint64_t',
+            'access' => 'readwrite',
+            'field' => "ctx->N",
+            'sanitycheck' => "value > 1 && is_power_of_two(value)"
+        ),
+
+        'KDF_PARAM_SCRYPT_R' => (
+            'type' => 'integer',
+            'ctype' => 'uint64_t',
+            'access' => 'readwrite',
+            'field' => "ctx->r",
+            'sanitycheck' => "value >= 1",
+        ),
+
+        'KDF_PARAM_SCRYPT_P' => (
+            'type' => 'integer',
+            'ctype' => 'uint64_t',
+            'access' => 'readwrite',
+            'field' => "ctx->p",
+            'sanitycheck' => "value >= 1",
+        ),
+
+        'KDF_PARAM_SCRYPT_MAXMEM' => (
+            'type' => 'integer',
+            'ctype' => 'uint64_t',
+            'access' => 'readwrite',
+            'field' => "ctx->maxmem_bytes",
+            'sanitycheck' => "value >= 1",
+        ),
+
+        'KDF_PARAM_PROPERTIES' => (
+            'type' => 'utf8_string',
+            'access' => 'readwrite',
+            'setaction' => qq(
+                if (!set_property_query(ctx, p->data) || !set_digest(ctx))
+                    return 0;
+            ),
+        ),
+
+        'KDF_PARAM_SIZE' => (
+            'type' => 'integer',
+            'ctype' => 'size_t',
+            'access' => 'readonly',
+            'field' => "SIZE_MAX",
+        ),
+    );
+);
+-}
+/* End of generated code */
+
+THe top level attributes are:
+
+- "name" is the name the functions will derive from e.g. "kdf_scrypt" to this will be appended _[gs]et[_ctx]_params
+- "functions" is the functions to generate.  By default both setters and getters but either can be omitted.
+- "prologue" defines some introductory code emited in the generated functions.  Function arguments are:
+  `void *vctx, OSSL_PARAM params[]` and this can be used to specialise the void pointer or declare locals.
+- "epilogue" defines some post decode code emited in the generated function
+- "params" defines the parameters both gettable and settable
+
+Within the "params" the fields specify each parameter by label.
+
+Each parameter is then specialised with attibutes:
+
+- "type" is the OSSL_PARAM type
```

Comment:
Already catered for :wink:

Specifically in this text:
- "code" set to "no" skips code generation for this parameter, it defaults to "yes" which generates handlers.  This is useful when a parameter is duplicated with differenting types (e.g. utf8 string and integer).

Reply:
Sooooo.....  does that mean that those cases will have to be done entirely manually?
Might I suggest another approach, with something like `type => [ 'string', 'int' ]`, and have the code generator cope with that?

For perl branching, if the `type` value is an array, `ref($whatever->[type})` will return `"ARRAY"`, and if it's just a scalar (like a string), it will return `""`

Reply:
Good suggestion.

Reply:
Thinking more about this and I suspect it runs into code generation issues.  Processing for int and string would be quite different, so there would have to be some demarcation between the two so separate code could be generated.  I'm not seeing how this would be less invasive than the code = no approach I was thinking of.

Still, good to keep it in mind as an option.